File: C++/Tree/Binary Search Tree Iterator.cpp
#include <bits/stdc++.h>
#include "TreeNode.h"

using namespace std;


/**
 * https://leetcode.com/problems/binary-search-tree-iterator/
 * */

class BSTIterator {
private:
    stack<TreeNode *> stack;

    void push(TreeNode *node) {
        while (node) {
            stack.push(node);
            node = node->left;
        }
    }

public:
    BSTIterator(TreeNode *root) {
        push(root);
    }

    int next() {
        TreeNode *curr = stack.top();
        stack.pop();
        push(curr->right);
        return curr->val;
    }

    bool hasNext() {
        return !stack.empty();
    }
};

File: Java-Kotlin/src/main/kotlin/tree/Binary Search Tree Iterator.kt
package tree

import java.util.*

/**
 * https://leetcode.com/problems/binary-search-tree-iterator/
 */

class BSTIterator(root: TreeNode?) {
    private val stack = Stack<TreeNode>()

    init {
        push(root)
    }

    private fun push(root: TreeNode?) {
        var node = root
        while (node != null) {
            stack.push(node)
            node = node.left
        }
    }

    fun next(): Int {
        val cur = stack.pop()
        push(cur.right)
        return cur.`val`
    }

    fun hasNext() = !stack.empty()

}
